### ISSUE TITLE
Understanding of the implementation of MPP loss 

### ISSUE THREAD

**Issue by lzx325** [lucidrains/vit-pytorch#124]
I am confused with the implementation MPP loss in `mpp.py`. In the original paper, assuming the number of channels is 3 and bits per channel is 3, the authors classify the mean color of the masked patch into a set of `2**(3*3) = 512` colors. In your implementation, however, it seems that you only classify them into `3*3*3 = 27` colors. 

Here is an exerpt of the code, with comments (1)-(7) added inline.
```{Python}
        avg_target = target.mean(dim=3)

        # (1) By default self.max_pixel_val is 1.0, self.output_channel_bits is 3; bin_size is therefore 0.333
        bin_size = self.max_pixel_val / self.output_channel_bits

        channel_bins = torch.arange(bin_size, self.max_pixel_val, bin_size).to(avg_target.device)

        # (2) By default, discretized_target is (batch_size, n_patches, n_channels)
        discretized_target = torch.bucketize(avg_target, channel_bins)

        # (3) Suppose n_channels=3, discretized_target is (batch_size, n_patches, n_channels, 3) = (batch_size, n_patches, 3, 3)` . 
        # (4) Values in the last dimension are mutually exclusive due to one-hot encoding.
        discretized_target = F.one_hot(discretized_target,
                                       self.output_channel_bits)
        c, bi = self.channels, self.output_channel_bits

        # (5) discretized_target is now (batch_size,n_patches,9). 
        # (6) The binary values of the last dimension are mutually exclusive in a group of three due to one-hot encoding.
        discretized_target = rearrange(discretized_target,
                                       "b n c bi -> b n (c bi)",
                                       c=c,
                                       bi=bi)

        bin_mask = 2**torch.arange(c * bi - 1, -1,
                                   -1).to(discretized_target.device,
                                          discretized_target.dtype)

        # (7) target_label can only have 3*3*3=27 distinct values
        target_label = torch.sum(bin_mask * discretized_target, -1)

        predicted_patches = predicted_patches[mask]
        target_label = target_label[mask]
        loss = F.cross_entropy(predicted_patches, target_label)
```
From the comment (1)-(7) above, I do not quite understand the reason to summarize the 27 distinct target label values into an integer that ranges from 0 (inclusive) to 512 (exclusive), because we are actually doing 27-class classification here. I am not sure if it is an implementation error or I misunderstand it. Thank you if you can clarify.

**Comment by lucidrains** [comment]
@zankner bringing in Zach, as he built the MPP module!

**Comment by lucidrains** [comment]
@zankner bringing in Zack, as he built the MPP module!

**Comment by lucidrains** [comment]
@zankner hmm, Zach wrote this piece of code so I'll let him work it out with you :)

**Comment by lzx325** [comment]
Dear Phil,
Thank you for contacting and hope we can hear from him soon.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by lucidrains** [comment]
@lzx325 I think I fixed MPP in 0.19.6 https://github.com/lucidrains/vit-pytorch/commit/64a2ef6462bde61db4dd8f0887ee71192b273692 , do you want to go over the code (and run it on your data) just to make sure?

@zankner sorry, it was just quicker this way :pray: you already did a great job :)

**Comment by LY080835** [comment]
@lzx325 Hello, what was the original mpp paper? Is it BEiT?

**Comment by LY080835** [comment]
or，What's the mathematical basis for mpp？
